### ISSUE TITLE
bad CPU type in executable: wine

### ISSUE THREAD

**Issue by kencu** [Gcenx/macOS_Wine_builds#3]
This is with the 5.11 Wine-Staging:
```
/Applications/Wine Staging.app/Contents/Resources/start/bin/winehelp: line 41: /Applications/Wine Staging.app/Contents/Resources/wine/bin/wine: Bad CPU type in executable
 Welcome to .

 In order to start a program:
   .exe: wine program.exe
   .msi: wine msiexec /i program.msi

 If you want to configure wine:
   wine winecfg

 To get information about app compatibility:
   appdb Program Name

% wine winecfg
zsh: bad CPU type in executable: wine
```
I guess your build machine must have a newer processor than this system I'm running.

**Comment by Gcenx** [comment]
I’m building on my mid 2014 13in MacBook Pro

I don’t believe that’s the issue here, your running macOS Catalina right?

**Comment by kencu** [comment]
yes, on a mid 2012 MacbookPro at the moment.

**Comment by kencu** [comment]
oh duh, I probably have to call wine64 instead of wine, don't I? if that iscall it is, sorry, blond moment....

**Comment by Gcenx** [comment]
Yeah you need to call wine64 over wine on Catalina.

Unless you use my wine-crossover build that includes a wine wrapper that launches wine32on64 when running Catalina and above.

**Comment by kencu** [comment]
thanks. I did want that, but had some trouble spotting it. I have it now.

You know a lot about wine!

**Comment by kencu** [comment]
that wine-crossover did it for me. The application I wanted (LTSpice) I guess has a 32bit installer. Now it works just great, and thanks ! See -- what goes around, comes around -- I helped a bit, now you helped me :>

BTW, I was seeing this:
```
% wine
./wine: line 29: /wine32on64: No such file or directory
./wine: line 29: exec: /wine32on64: cannot execute: No such file or directory
```
and I looked at the script, and added this:
```
export WINESKINPATH="/Applications/Wine Crossover.app/Contents/Resources/wine/bin"
```
and after that it works for me. No doubt I did something else wrong somehow to cause me to need to do that, but at any rate, perfectly running wine on Catalina now.

**Comment by Gcenx** [comment]
I could have sworn I’d uploaded the fixed version that does this already.

I probably only uploaded the fixed strip for Wineskin but not the cask...

I wonder if it’s time to look into handling the CodeWeavers version of llvm/clang-8 and updating the wine-crossover Portfile so it can be built directly using the Portfile.

**Comment by kencu** [comment]
I am the maintainer of llvm/clang on MacPorts, if that is of any help to you :>

**Comment by Gcenx** [comment]
Yep I didn't upload a fixed version, the fixed Script was only within the newer Wineskin wrapper I'm working on. I'll uploaded it once compressed since it also includes `DYLD_FALLBACK_LIBRARY_PATH` to avoid some dylibs not being found still.

> I am the maintainer of llvm/clang on MacPorts, if that is of any help to you :>

I've considered it, the version of llvm/clang-8 from CodeWeavers isn't an entire compiler suite.
However I had asked qparis to create a fully working version for Cross-Compiling, he provided a patch set to he did apply to the debian llvm/clang-8 the patch can be found [Here](https://github.com/PhoenicisOrg/phoenicis-winebuild/blob/master/environments/ms_hook_prologue/0003-Add-Wine32-Support-Full.patch)

I was considering just making an individual Port that would build the custom llvm/clang-8 and name them differently then the mp versions, as the only real use for this custom llvm/clang-8 currently is compiling wine(64/32on64)

**Comment by kencu** [comment]
There are other examples of this too. This port installs a special version of clang/llvm for it's own use with a custom configuration. It's probably a good example for you. <https://github.com/macports/macports-ports/blob/master/lang/ispc/Portfile>

**Comment by kencu** [comment]
BTW -- this is just fantastic. The version of LTSpice that runs under wine is from last week, and works superbly well. It even accesses all the files I made with the Mac version, opens, runs them, and saves them back into my (icloud-sync'd) drive without the slightest bit of me having to force it to do anything.

Wonderful. I know wine is responsible for a lot of this, but your packaging is great, and makes it all work and without you doing it, we'd be nowhere.

**Comment by Gcenx** [comment]
I’m glad you’ve found the package(s) useful

The packaging was partly based on Winehqs app bundle, but I’d opted to follow KISS over forcing XQuartz onto people like the old Winehq packages that were missing key dependencies.

The `wine` wrapper was added due to KISS and ensuring winetricks is compatible over forcing a patched version like I did prior.

Side Note;\
Before working on the needed custom llvm/clang-8 Portfile I’ve tackled another useful one that can’t be added to macports-ports directly sadly, the macOSX.sdk that Ryan was creating that needed Xcode to be pre-downloaded.

Instead I’ve opted to directly download the SDKs from another location then symlink them into `/Library/Developer/CommandLineTools/SDKs/`

For the custom llvm/clang-8 I don’t know if that could be accepted into macports-ports so I might just make a Portfile that downloads a pre-built Xcode Toolchain package instead

**Comment by kencu** [comment]
I asked Ryan about it -- he says he has made up his own port for the modified clang/llvm toolchain but not released it yet.

**Comment by Gcenx** [comment]
Did he mention why he didn’t publish it somewhere as I’ve had to help a number of people compiling it along with compiling wine32on64

**Comment by kencu** [comment]
I think just finishing it up with his wine update he has planned, and will roll it all out together.

**Comment by Gcenx** [comment]
@kencu question does libtapi support tapi-tbd-v3 currently as I only see tapi-tbd-v2 listed

**Comment by kencu** [comment]
It does, I believe. I used the latest available <https://opensource.apple.com/source/tapi/> when I put the port together.

**Comment by Gcenx** [comment]
I know osxcross definitely does but haven’t required it on macports as of yet but might for some other Portfiles I’ve been working on.

The only notable one will be `wine-staging` definitely no longer compiles using the 10.8 SDK and lower but I can simply resolve that with my version of the MacOSX.sdk Portfile (10.4u>10.15)

I figure what’s taking Ryan so long with updating the wine-* ports, he’d be better waiting on CrossOver-20 as adapting the current wine32on64 patches isn’t worth the effort.

**Comment by Gcenx** [comment]
I have a weirder question, is is possible to for e a Portfile to skip the linker checks?

**Comment by kencu** [comment]
We can probably make it do most anything if motivated enough. Show me the exact thing you need to do and I'll see what comes to mind.

And one back -- I have patched wine 4.04 to run on 10.6.8 just because, but I need to compile a new "mono" for it as the one I can download generates errors. Is that hard to do?

**Comment by Gcenx** [comment]
On installation of the older macOSX.sdk files macports does its usual checking for linker errors then causes macports to wrongly believe the SDK is damaged.

Honestly wine-mono has been mostly useless until very recently, but complaining it shouldn’t be an issue https://github.com/madewokherd/wine-mono just need the usual mono tools and mingw to compile it.\
I’d recommend skipping wine-mono and instead just installing dotnet45 using winetricks.

I’ve not done much testing of Wine on pre-10.9 systems due to the tls requirements being higher now meaning access GitHub is more annoying, but I’d assume we could still build the latest versions by using a newer SDK might be required.

**Comment by kencu** [comment]
Making an SDK-driven build use a different linker than the one in the SDK is hard unless the build system allows a method for overriding parts like "LD". There is an SDK-overlay method for Apple SDKs that I have not fully sorted out with MacPorts builds as yet.

For practical purposes, for one-off builds, I will often just symlink the tool I want into the SDK and then put the old one back later.

**Comment by Gcenx** [comment]
Seems I’m not explaining clearly.

When installing a Portfile, it goes through multiple stages, downloading distfile, checksums, extraction, configure, build, destroot.

After the above is completed macports then checks all installed dylibs to ensure they are linked correctly, the problem is the dylib files provided in older SDK files want to check system locations instead of the SDKROOT that’s expected.

Let’s say;\
MacOSX10.12.sdk/System/Library/Frameworks/OpenCL.framework/Versions/A/Libraries/AMDil_r700.dylib\
Is expecting to find a dylibs within the SDK that’s no precent in current versions of macOS but is contained within the SDK files, macports will freak

If needed I can provide terminal output to better explain the issue.

**Comment by kencu** [comment]
Oh, I see. Yes, MacPorts will run all the executables and dylibs to check linkages after installing the port. These libraries need to be installed in the system.

During the build, the build system sets the "root" to the SDK, eg "MacOSX10.12.sdk" . The links generated during the build are  relative to that root, in your case "/System/Library/Frameworks/OpenCL.framework/Versions/A/Libraries/AMDil_r700.dylib".

But on installation, there needs to be an actual library at "/System/Library/Frameworks/OpenCL.framework/Versions/A/Libraries/AMDil_r700.dylib" for the installed software to link against. It can't be in the SDK at that point, it has to be really there in the root "/" somewhere.

That's not really a macports thing (although you can turn off that link check if you want by manipulating this variable in macports.conf:
```
revupgrade_autorun
Controls whether the rev-upgrade action will be run automatically after upgrading ports.

Default: yes
```
or by passing a flag to port - "--no-rev-upgrade" I think it is.

But -- why? The software you are installing is broken if it fails rev-upgrade, and won't run -- it will just error out with a message that the library is not found.

**Comment by kencu** [comment]
you probably know about install_name_tool, that lets you manipulate these links. I sometimes have to do that to force things around if needed, eg this bit of hackery <https://github.com/kencu/TigerPorts/blob/cfc9cda936af95b333e36c6ad5abc48769d86e91/lang/llvm-7.0/Portfile#L742>.

**Comment by Gcenx** [comment]
--no-rev-upgrade allowed the MacOSX10.10.sdk and lower to install without macports complaining but that’s not ideal. I was hoping there was some other hidden option that could be added to the Portfile to skip this check, but I guess this wasn’t exactly something that was considered before.

As these are actual MacOSX.sdk files install_name_tool is out of the question else the resulting compilation using the installed SDK will result in a broken compile.

It looks like the lowest version that can be used then would be MacOSX10.11.sdk by just removing the mentioned AMDil_r700.dylib as that’s not even a real dylib but something else that’s used in the system by ATI graphics cards. (I did try turning it into a tbd stub)

This means MacOSX10.11.sdk is the lowest SDK that can be installed using a Portfile

I guess I can just have wine-staging install and use the MacOSX10.11.sdk (along with ld64 ect) and set the min version low enough it shouldn’t be an issue.

**Comment by kencu** [comment]
MacPorts will never allow Portfiles to set no-rev-upgrade as they would install broken software. It's only useful for devs.

install_name_tool is your fix, if it is fixable at all. I would still need to see the actual problem you're having rather than this discussion, but consider:

You build your software against the MacOSX10.12.sdk and it links against some dylib. You install it and that dylib doesn't exist in the system you've installed to. It's totally broken and useless. You have to come up with a different method.

If the dylib you're linking against doesn't exist in the system you're installing to, you need a different solution -- don't use that dylib, use a similar one in a different place, install your own and link against that, blah blah blah. But linking against an SDK and expecting that to be used in the end in a non-starter. Most of the dylibs in the current SDKs have no object code in them anyway -- they are just text files with the names of routines -- that is what TAPI is all about.

**Comment by Gcenx** [comment]
Maybe I’m not explaining this well enough.

We can currently use MacOSX10.13.sdk and compile software to run on 10.6 (as long as we provide libraries that macOS would be missing)

For example the wine-crossover you installed was compiled on my Mojave system using the MacOSX10.13.sdk for all libraries (macports +patches) and a modified MacOSX10.14.sdk that allows i386 target (wine respects dark mode on Mojave)

**Comment by Gcenx** [comment]
That wine-crossover build also works on OSX10.9 it could also work on lower versions but with the recent tls requirements I’d skipped going below 10.9 for the moment.

**Comment by kencu** [comment]
> Maybe I’m not explaining this well enough.

I think if you need me to help any deeper I'd need to see the exact problem, agreed:)

Re: TLS -- if the software can build against gnutls or openssl or similar, all is good; if it can only use the Apple Security framework, Apple TLS, etc, then you're hosed. Most unix-y things do use gnutls with a switch if asked...thank goodness.

**Comment by Gcenx** [comment]
I've committed what's currently working. (MacOSX10.10 and lower were stripped)

Seems I'm unable to compile wine-5.13 using the MacOSX10.9.sdk due to missing symbols, also attempted to use the MacOSX10.13.sdk and failed due to ld64 not supporting TAPIv3 only TAPIv2.\
Compiling using MacOSX10.11.sdk on 10.9 fails something with the linker not finding the SDK provided files it looked like.

The current minimum SDK that could compile wine now was MacOSX10.10.sdk with the additional workaround I'd added to disable the Metal header detection

**Comment by kencu** [comment]
make sure that you're using the right ld64, of course.

you want to install ```ld64-latest``` and then set up the symlinks to use it.

and then you might need to force your build to use macports' ld64 instead of the ld in the SDK, which sometimes takes some extra work.

**Comment by kencu** [comment]
macports ld64-latest with our current libtapi supports tapi v3 <https://opensource.apple.com/source/tapi/tapi-1000.10.8/include/tapi/Core/TextStub_v3.h.auto.html>.

**Comment by Gcenx** [comment]
Ah so `ld64-latest` is what I’ll be needing then, I was thinking it’s weird that osxcross supported TAPIv3 but macports didn’t.

Did a little digging for what wine-devel-5.13 seems to want and it seems your legacysuppory should cover the requirement(s) I've currently encountered issues with.

As the compile is running from within a VM this might take a while to confirm but it’s gotten past the previous issues already so fingers crossed

**Comment by kencu** [comment]
And you're convincing me about what I already felt was the case -- we have to simplify our ld64 port. People should get a clear message about how to use it, and obviously we are failing in that regard. This one is on us.

**Comment by Gcenx** [comment]
Well I have some news on using legacysupport, it allowed wine to compile, but it doesn’t work somehow the wine-preloader mapping got screwed up.

I’ll check your example from before for forcing macports ld64 to hopefully allow compiling without needing to patch wine heavily.

Still need to test building using a 10.10 VM to see if it does build natively as it did build on Mojave using the 10.10 SDK when setting the Metal header check to disabled

**Comment by kencu** [comment]
I know nothing about wine preloader, but FYI I added an option to legacysupport to link in the library statically; it's possible you might find that works better for some purposes: <https://github.com/macports/macports-ports/blob/8147397edb6d30c44d172b3e1233b732eaf059f6/_resources/port1.0/group/legacysupport-1.1.tcl#L32>

**Comment by Gcenx** [comment]
This took longer then I’d have liked due to running within a VM.

OSX 10.9 and above will are now compiling wine-devel-5.13.

With the need to workaround the the newer tls requirements I didn’t really get to do much testing within the 10.7 VM, it compiles and launches winecfg/wine explorer but macdriver doesn’t work with recent wine versions within a VM meaning testing is very limited

**Comment by Gcenx** [comment]
@kencu could you please check your able to compile `graphene` from source and it installs `/opt/local/include/graphene-1.0/graphene-gobject.h`as I just rebuilt my environment and this header was not precent causing  `gstreamer1-gst-plugins-base` to fail

I believe the recent changes to meson caused this issue as I rebuilt my environment last month using my overlay without issue

**Comment by Gcenx** [comment]
I believe the ObjectiveC part of the compile never happens, I believe the crossfile need to have the following section added for this to work;

```
[binaries]
objc = '-target'
```

I’ll confirm this once I’m done with work

**Comment by kencu** [comment]
I was hoping to leave all the binaries out of the crossfiles -- as we then get into a real mess of compiler specs.

Maybe your build is failing for the other 102,000 reasons builds fail :>

If I have to add the binaries spec, it will have to be custom-written for each build using MacPorts selected compiler for that build, as a second layered crossfile into the build dir. That is a total UGH.

**Comment by Gcenx** [comment]
I kept digging since the above comments and I believe it’s just that meson when using the crossfile isn’t detecting objc binary or env variable.

It seems when meson is in cross mode it skips the usual check it makes to see if the compiler supports ObjectiveC, maybe placing a symlink to the selected compiler during build or setting some env will resolve the issue without needing to add additional items into the crossfiles

**Comment by Gcenx** [comment]
@kencu after doing multiple builds and looking over while running multiple `graphene +universal` builds using verbose I can see the problem, after swapping to using crossfiles meson no longer finds anything.

Usually `graphene` needs both g-ir-scanner & g-ir-compiler, but using a crossfile these binaries are no longer found, this also includes cmake, pkgconfig or any other binary it wants.

I did managed to get `graphene` working again at least by adding only one section

```
[binaries]
pkgconfig = '/opt/local/bin/pkg-config'
```

With only pkgconfig added `graphene` found both g-ir-scanner & g-ir-compiler meaning the compile now includes the needed support plus the additional header.

**Comment by kencu** [comment]
Thanks for finding this and sorting this out. I will reproduce, and then open a ticket and see what the best way forward here might be. The meson PortGroup could write a custom binary crossfile for each build based on the current portfile specs -- it layers on top of the machine ones already there -- I was just hoping that we wouldn't have to do that. But perhaps it's inevitable .The qt5 portgroup already does something like this so there are models to follow. Thanks again!

**Comment by Gcenx** [comment]
@kencu this particular one took me a while to notice since `graphene` was built before the meson changes.

I’ve added the workaround to macports-wine along with some others (cario|cario-devel suppress i386 linker warnings for 10.13 and greater)

I believe most of the workarounds are at least clean enough now bar the FAudio one, I’m using curl to workaround GitHub a TLS requirements with an additional checksum section, I’d like to instead grab the same archive that GitHub group gets so another checksum section is unneeded.

**Comment by kencu** [comment]
If the TLS issue is with macports downloading things, we wouldn't put individual fixes into each port for that. We just build macports against a newer curl instead. There is a macports configure arg for that, and then everything works.

I'm still updating the 10.13 universal system to see what's up with graphene at the moment ..

**Comment by kencu** [comment]
graphene builds +universal for me with the current meson and current crossfiles without any modifications.
[graphene-build-universal.txt.zip](https://github.com/Gcenx/macOS_Wine_builds/files/5057721/graphene-build-universal.txt.zip)

**Comment by Gcenx** [comment]
@kencu it will build without error however now try to build gstreamer1-gst-plugin-base and it will fail due to a missing header graphene-gobject.h

**Comment by kencu** [comment]
Right.... OK, let me see about that.
```
Found Pkg-config: NO
Found CMake: NO
Run-time dependency gobject-2.0 found: NO (tried pkgconfig, framework and cmake)
```

**Comment by Gcenx** [comment]
> but be careful adding too many "fixes" to your ports as you may have to undo them all shortly...

Once I’m aware a workaround is no longer required I remove them

**Comment by Gcenx** [comment]
I believe your referring to https://trac.macports.org/ticket/51516 yeah I’m hopeful once this happens I won’t need to deal with workarounds for downloading distfiles

**Comment by kencu** [comment]
If you like, just reinstall macports from source, specifying a new curl, right on top of the existing macports. It takes three or four minutes.

If you have a separate curl somewhere, use it, otherwise you can self-reference if you stay a bit careful. Assuming you have installed ```curl``` in ```/opt/local```, do this to configure macports:
```
configure --with-curlprefix=/opt/local
make
sudo make install
```

I install curl into a macports installed in ```/opt/bootstrap``` and reference that, but that's 'cause I'm paranoid :>

**Comment by kencu** [comment]
It looks like meson, when cross-building, will not just search the path for pkg-config. It wants it to be specified in the environment. I added this to the graphene portfile:
```
configure.env-append PKG_CONFIG=${prefix}/bin/pkg-config
```
and that made ```graphene``` happy:
```
Using 'PKG_CONFIG' from environment with value: '/opt/local/bin/pkg-config'
Found pkg-config: /opt/local/bin/pkg-config (0.29.2)
Run-time dependency gobject-2.0 found: YES 2.58.3
```

1. How to handle this in the general case, rather than adding that spec to every portfile?
2. What other binaries might need the same treatment? Stay tuned!

**Comment by kencu** [comment]
Just for laughs, meson seems perfectly happy finding other binaries in the PATH:
```
Build-time dependency gobject-introspection-1.0 found: YES 1.60.2
Program g_ir_scanner found: YES (/opt/local/bin/g-ir-scanner)
Program g_ir_compiler found: YES (/opt/local/bin/g-ir-compiler)
Build targets in project: 3
```

**Comment by Gcenx** [comment]
@kencu I used your instructions on some VMs without issue but for some users they struggled so workarounds lol.

While you can do that for each Portfile I instead modified the meson portgroup to required pkgconfig and meson crossfile now includes pkgconfig the prefix is resolved when installed check macports-wine to see what I did

**Comment by Gcenx** [comment]
@kencu I assumed macports-base already handled PATH....?

**Comment by kencu** [comment]
The crossfile doesn't need touching. I don't think. I didn't touch it. 

Just meson wants PKG_CONFIG spec'd when cross building.

**Comment by kencu** [comment]
base sets PATH -- but meson seems to ignore for the specific search when looking for pkg-config when cross-building. 

Which makes sense.

**Comment by Gcenx** [comment]
Not really most cross environments will search PATH but look for tools for the specified target, so for darwin it will be searching for all tools appended with darwin

**Comment by kencu** [comment]
I'm trying to imagine if there is some port that wants to set it's own PKG_CONFIG (like some cross-building compiler) to something other than the default one, that builds with meson, and would be messed up if we overwrote that in the meson PG.

That's a stretch.

There are dozens of ports that build with meson that don't use pkg-config though. How will it work when we spec a pkg-config in the environment that doesn't actually exist? That is ugly programming, but likely no problem in the end, I guess. That is EXACTLY the kind of stuff the MacPorts admins detest though, and for good reason.

**Comment by kencu** [comment]
meson is clearly not searching the PATH for pkg-config when cross building, otherwise it would have found it without the spec.

**Comment by Gcenx** [comment]
Can the meson portgroup handle passing PATH when using cross file?

**Comment by kencu** [comment]
OK - I see what you mean -- maybe meson is searching the path for some kind of frankenstein binary named ```i386-pkg-config-darwin```

meson. new birthing build systems can be such a PITA.

**Comment by kencu** [comment]
> Can the meson portgroup handle passing PATH when using cross file?

base is already setting the PATH correctly, eg. above:
```
Program g_ir_scanner found: YES (/opt/local/bin/g-ir-scanner)
Program g_ir_compiler found: YES (/opt/local/bin/g-ir-compiler)
```
but not for ```pkg-config```.

**Comment by Gcenx** [comment]
I meant some more correct way to have it detect pkg-config, I didn’t try making the pkg-config wrappers but that might work

**Comment by kencu** [comment]
It's easier to set this:
```
configure.env-append PKG_CONFIG=${prefix}/bin/pkg-config
```
than to start adding wrappers for everything.

**Comment by kencu** [comment]
If wrappers would even work (we have no idea if they would or not). But even if they did, I wouldn't use that.

We'll just add PKG_CONFIG to the env, along with all the other things we already add.

**Comment by kencu** [comment]
The way ```meson``` searches for ```pkg-config``` is specified here <https://github.com/mesonbuild/meson/blob/847d652c4ef179bbf94c0e322dc3277858fcf08c/mesonbuild/dependencies/base.py#L557>

**Comment by Gcenx** [comment]
Once a more correct way it within macports-ports I’ll remove the while working janky workaround.

Yeah I’d avoid wrappers when possible, it’s get annoying after a while to maintain them

**Comment by kencu** [comment]
Opened: https://trac.macports.org/ticket/60987 to get some feedback before we decide whether to fix this in base, the meson PG, or in each portfile individually.

Thanks again for noticing! You're really helping me kick the tires here, by pushing MacPorts to it's limits!

**Comment by Gcenx** [comment]
@kencu i think you might be interested in this https://github.com/tpoechtrager/apple-libtapi

This repository is used by https://github.com/tpoechtrager/osxcross and constantly tried to support the latest SDK versions.

According to the repository it should be possible to build libtapi on MacOSX10.9

**Comment by kencu** [comment]
I'm the maintainer of libtapi on macports, and fixed it back to 10.6. could probably do 10.4 and 10.5 too, but no need as yet.

This lets the very latest ld64 run on 10.6 up...!

**Comment by Gcenx** [comment]
Ah seems I misread the comments within ld64

**Comment by Gcenx** [comment]
i

**Comment by Gcenx** [comment]
Ah it seems I’d misread the comments within ld64, it’s not that macports can’t build the newer tapi it’s that some Xcode versions can’t...

**Comment by kencu** [comment]
libtapi needs a fairly new clang to build, but I worked all that out I thought. It gets green lights across the oard <https://ports.macports.org/port/libtapi/summary>.

Is there something specific you're hoping to do with libtapi or ld64 that I can help you with?

**Comment by Gcenx** [comment]
The last time I’d attempted to build using the MacOSX10.13.sdk on 10.8/10.9 it failed to read the tapi v3 stubs so I needed to fallback to MacOSX10.11.sdk instead.

I’d guessed the prebuilt copy of tapi was outdated as for 10.8/10.9 testing I avoid building clang from source due to it being a VM

**Comment by kencu** [comment]
there is a bootstrapping issue with ld64 that presently needs a manual step to overcome.

on many systems you can't build ld64-latest directly with the system xcode.

You have to install the newest ld64 that the os can build (different for each)' then install clang-9.0 and cctools, and then you must manually upgrade ld64 to ld64-latest. It takes 15 seconds. 

I'd love to automate this, and have some ideas, but even the simplest fixes I have are getting pounded at MP right now. We have an open ticket somewhere about this issue.

**Comment by Gcenx** [comment]
Yeah I’ve seen numerous open issues that seem like they could have been resolved long ago.

I’d rather avoid as many manual steps as possible to avoid user error when possible (that’s what most of the modifications on macports-wine is all about, same for the Winehq packages here and the brew tap)

**Comment by kencu** [comment]
> I’d rather avoid as many manual steps as possible to avoid user error when possible (that’s what most of the modifications on macports-wine is all about, same for the Winehq packages here and the brew tap)

Me too. If you can think of a robust way to bootstrap ld64 all the way to ld64 latest while obeying MacPorts rules and regulations in the process, I'm all ears.

**Comment by Gcenx** [comment]
> Me too. If you can think of a robust way to bootstrap ld64 all the way to ld64 latest while obeying MacPorts rules and regulations in the process, I'm all ears.

I’ll look up the ticket and see what I make of the current discussion.

**Comment by Gcenx** [comment]
@kencu I haven’t found the ticket your referring to, but I have noticed some odd behavior within ld64/ld64-latest it’s requiring macports-clang9 this is rather a heavy requirement when it’s possible to build tapi/ctools/ld64 with clang3.9+ outside of macports, I wanted to ask before opening a ticket

**Comment by kencu** [comment]
our libtapi is set to require the same, clang-3.9+ and xcode clang > 800.

<https://github.com/macports/macports-ports/blob/8ce58421375ec75b929bc222ac9ea7193bd1c952/devel/libtapi/Portfile#L67>

macports compiler defaulting does not call for the least capable compiler that can work, though. If the system compilers are incapable, it goes to clang 9 first (as it should).

**Comment by kencu** [comment]
even if you set libtapi to build with clang-5.0 (oldest we have that can do it) you're no further ahead.

what is needed is an automatic stepping stone, where a throwaway ld64 is used temporarily to build ld64-latest. The problem is that macports is not wired up that way, and ld64 is wired into all the compilers directly.

Possibly I could disable one of them like clang-5.0 to use the ld64 that is the best the os can make, and the use that to build ld64-latest, and then wire to into later clangs...that won't be accepted well though.

Maybe I could use a custom build of clang and ld64 to generate libtapi and ld64-latest...

But--these are all moving targets, and what builds libtapi today may not build the next version of libtapi tomorrow, and that's a lot of work for a collapsing house of cards...

**Comment by Gcenx** [comment]
What I’m trying to do is install ld64-latest without requiring macport-clang unless it’s required to build it.

I believe the issue is the way I’m intending to use ld64 is out of scope of macports

**Comment by kencu** [comment]
any system that comes with xcode clang 900+ can do it. That is what MacPorts ld64 port is already set up to do, though, so you don't have to do any work to make that happen.

We use ld64_xcode where that is newer than ld64-latest.

Maybe better to tell me exactly what your problem is?

**Comment by kencu** [comment]
actually, it needs xcode clang >801 .

**Comment by Gcenx** [comment]
The issue in having is I want to install ld64-latest without macport-clang-9 also being installed.

The idea being when I’m running my build let I’ll force ld64-latest so there are no irritating i386 warnings displayed

**Comment by kencu** [comment]
I getcha. The logic to decide that is <https://github.com/macports/macports-ports/blob/b0aafd8b14d38d040ebf92787cedeb29543e9aa2/devel/ld64/Portfile#L227>.

Your system wants to use ld64_xcode, and that is generating the error you mention? That only happens if you are running darwin15, or have xcode 11+ installed, for reasons explained in the portfile at that link area.

You can change it though. It's easy to force that ld64 Portfile to install ld64-latest on all systems.

**Comment by Gcenx** [comment]
For some reason this section seems to be getting called https://github.com/macports/macports-ports/blob/b0aafd8b14d38d040ebf92787cedeb29543e9aa2/devel/ld64/Portfile#L308

So that’s forcing +llvm90 on my system when I’d guess that should be the case

**Comment by kencu** [comment]
Well, we do have ld64-latest set to build with llvm90 by default, as it's the practical newest llvm.

How is that bothering you? The way this is supposed to work is that you, on 10.14, ask for "ld64".

The logic sorts out that ld64 for you should be `ld64-latest +llvm90`.

And your xcode toolchain builds llvm90, and then ld64-latest using it, and sets that to be your default.

Unless you purposefully try to override the variants, or (here is a gotcha sometimes), unless you're upgrading from some older setup and the variants want to propagate on you.

**Comment by Gcenx** [comment]
That’s the thing I want `ld64-latest` without +llvm90

Other then the Xcode provided version of llvm the other custom one I require is already bundled within an XCode toolchain

**Comment by kencu** [comment]
For that then, you'd have to make up your own custom version of ```ld64```, something like ```ld64-latest +llvm_xcode```, which we don't (and won't) offer, and then change this line:

<https://github.com/macports/macports-ports/blob/b0aafd8b14d38d040ebf92787cedeb29543e9aa2/devel/ld64/Portfile#L471>

to point to the llvm-config in xcode somewhere that you want to use.

**Comment by Gcenx** [comment]
I was more hoping something like `ld64-latest -llvm90` could be called then overriding the Portfile.

Guess I need to look into an alternative approach for this one.

For the moment the osxcross version of ld should be enough

**Comment by kencu** [comment]
that is built against xcode's llvm? I'd be surprised if so -- it is a linux primarily, and I think it also builds against it's own installed llvm.

**Comment by Gcenx** [comment]
Yes it’s built against Xcode-llvm osxcross does support the setup.

I’d setup osxcross precisely on macOS to speed up my development on phoenicis-winebuild.

Since I have macports installed and patched to build +universal i now symlink to my real macports install.

To make my life easier thrown together osxcross Portfile that disabled building the dependencies and instead uses the versions from macports like libtapi

**Comment by Gcenx** [comment]
Also no by default osxcross doesn’t built its own compiler unless you request it to build llvm or gcc

By default the system clang is used with wrapper binaries that inject sdkroot to clang and by setting the correct host setting it will default to using the osxcross binary wrapper for everything (ld isn’t a wrapper)

**Comment by Gcenx** [comment]
osxcross when built using Xcode ld doesn’t support LTO (not ideal)

I really don’t want to mess with ld64 or related ports and having macports-clang installed won’t be beneficial (I have any needed llvm/clang version as Xcode toolchains)

The only alternative is can think of is patching all relevant sources and attempting to build an XCode toolchain and fix rpaths to be relative if needed

**Comment by kencu** [comment]
I'm sorry about that -- the ld64 port uses llvm-config to get it's configuration from; I don't think there is a similar thing installed with xcode. It is probably possible to hack into the port to make it build against the llvm installed with xcode, but -- too much work for me.

You know you don't need the entire clang toolchain for ld64 to work; just llvm, any version (defaults to 9.0) so that ld64 can do the LTO.

If all you are trying to do is suppress the ld64 warning, perhaps you might consider a tiny wrapper shell script that adds ```-w``` to the link line and then calls the xcode one. There are lots of simple examples of how that would work.

**Comment by kencu** [comment]
In fact, you might just hack the existing ```ld64-xcode``` script to make it do that ;>

**Comment by Gcenx** [comment]
I was precisely injecting -w but that’s far from ideal, a wrapper script adding -w installed into PATH would cause even more issues.

Around 2 months back I’d removed all but two wrapper scripts from phoenicis-winebuild, the remaining two are just to fake i386-apple-darwin17-gcc and x86_64-apple-darwin17-gcc, now setting --host x86_64-apple-darwin17 is all that needs to be set since osxcross then injects includes/headers/frameworks.

I’d look into getting the Portfile to support building using Xcode llvm just without LTO but I’m assuming that’s another edge case

**Comment by kencu** [comment]
I don't know why you'd say adding ```-w``` to the existing wrapper script would be so bad, but of course that's up to you.

If you're going to use ld64 +llvmNN you might as well use LTO, as it is just sitting there waiting to be used anyway. Otherwise just build ld64 without LLVM (and without LTO) and you're in the same position.

And why you find it so bad to install llvm-9.0 I'm not sure either; that just works, no more hiccups. I'm going to make ld64 default to llvm-10 soon on suitable systems, and llvm-11 is rc2 that will be coming shortly as well.

**Comment by Gcenx** [comment]
Adding -w to the ld64-Xcode script would mean no more linker warnings, cario/cario-devel only builds i386 on Mojave with -w being passed dbus it’s not the i386 warning that’s causing the issue but something with the configure checks and the version of binutils shipping within XCode10+

The idea was to have a recent enough version of ld64 that doesn’t add the i386 linker warnings but still supports newer SDKs, having LTO or not didn’t really matter in my case.

I know it seems strange but I’d prefer to not install llvm/clang into the environment if at all possible since I have XCode toolchains configured with each llvm version needed include CodeWeavers-llvm-8.0.0

**Comment by Gcenx** [comment]
There I believe I opened all required tickets to cover everything properly bar Apple LLVM/Clang to support Apple silicon from current Intel systems.

ld64-latest, libtapi, cctools and xar (swap a fork)

**Comment by kencu** [comment]
heh -- now i just have to spend a month doing it all :>

**Comment by Gcenx** [comment]
It might take you a while but it will be the macports way, now if did updated those it would lightly only work on a couple of configurations

**Comment by Gcenx** [comment]
xar is already done as a patch

**Comment by Gcenx** [comment]
@kencu I tried the WIP Portfile you provided for libtapi and had the issue you mentioned in the commit, patching that didn’t resolve it.

Hell even using apple-libtapi repository didn’t resolve the issue, so it looks like the workarounds used within build.sh is required to get it compiling.

Not had chance to test that yet with CrossOver-20-Beta1 being available

**Comment by Gcenx** [comment]
@kencu i didn’t directly cc for this but since it’s related gcc-10.2.0 maybe you might have some input https://trac.macports.org/ticket/61251

**Comment by Gcenx** [comment]
@kencu I'm sure you'll notice but you might be interested in https://github.com/macports/macports-base/pull/213

**Comment by kencu** [comment]
well isn't that neat...who knew they would still have the i386 bits in those headers?

I will be interested to see if macports will allow this in at this point, but for those of us interested, it's indeed useful.

**Comment by Gcenx** [comment]
I thought everyone know about this but Macports dropped 32Bit anyway as the SDK didn’t support it anymore (without patching)

I don’t think it will be accepted at least in its current form but maybe another Macports member will implement it in a more Macports friendly manner

**Comment by lpla** [comment]
Would it make sense to replace `/Applications/Wine\ Stable.app/Contents/Resources/wine/bin/wine` and `/Applications/Wine\ Stable.app/Contents/Resources/wine/bin/wine-preloader` binaries with the 64-bit equivalent now that it is not possible to run 32-bit .exe files in Catalina/Big Sur? This would avoid issues when opening 64-bit .exe files from Finder and related scrips or apps like `winehelp` in modern macOS versions without 32-bit support.

I guess that a separate package should be done for this as older macOS versions still work with the current wine install. Am I too naive?

**Comment by Gcenx** [comment]
The packages are built to be replacements for the now depreciated Winehq packages.

The packages containers both wine and wine64 binaries there’s also no 64Bit equivalent of wine outside of crossover based sources aka wine32on64 (must be that exact name)

The scripts could be tweaked more if needed.

The binary wrapper that launches Terminal and allows launches via Finder could be modified more or edit the plist.info to use a 32Bit version below Catalina and use a 64Bit binary when above.

As for 32Bit exe files you can can run those by using my wine-crossover package available from my brew tap [gcenx/wine](https://github.com/Gcenx/homebrew-wine)

**Comment by kencu** [comment]
I have been using the wine crossover package on Catalina since you released it, and it seems to work fine to run 32bit apps.

The one I really wanted was LTSpiceVII, and that has a 32bit installer it appears.

And -- again -- thanks for this!

**Comment by Gcenx** [comment]
@kencu I recently found something weird bug with `gmp` when built as +universal the i386 slice is working just fine but the x86_64 slice doesn't work. Yet a the i386 slice on it's own isn't and issue and the macports provided x86_64 one also works just fine?

This was tested with MacOSX10.13.sdk and also "/" both produced the same results.

I manually run lipo on the single arch builds of gmp and the resulting merged dylib both slices work correctly.

**Comment by kencu** [comment]
Universal building is a bit tricky, and with something like gmp where there is a lot of assembly, the chances for an error are higher. If you get to the point where it is a distinct error, please open a ticket and we can work through what is going on.

**Comment by Gcenx** [comment]
Thanks @kencu I’ll see if I’m able to find a distinct error then open a bug report.

**Comment by Gcenx** [comment]
@kencu for my purposes I can say I’ve resolved this. As I need to use the same dylibs across multiple systems forcing core2 and having assembly disabled resolved the issue I was having, need to retest if this also works with assembly enabled as I need these to also function under Rosetta2.

Are the prebuilt x86_64 gmp packages (Intel) built with assembly enabled?, I know it used core2 on the buildbot and the prebuilt package worked no issue under Rosetta2.

**Comment by kencu** [comment]
looks like assembly is enabled for all but arm64 at present. There is a PR to enable arm64 assembly.

**Comment by Gcenx** [comment]
@kencu needed to add another new dependency for the upcoming wine-6.2 [jxrlib](https://github.com/Gcenx/macports-wine/tree/master/graphics/jxrlib), I’m sure that’s not the Macports way but it works.

The makefile is useless due to no install target & doesn’t produce dylibs that are expected by almost everything on macOS including wine lol.
